File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // hashmap + sliding window (dynamic)

        int left = 0;
        Map<Character,Integer> charmap = new HashMap<>();
        int longest = 0;


        for (int right = 0; right < s.length(); right++)
        {
            charmap.put(s.charAt(right), charmap.getOrDefault(s.charAt(right), 0) + 1);

            while (charmap.get(s.charAt(right)) > 1)
            {
                 charmap.put(s.charAt(left), charmap.getOrDefault(s.charAt(left), 0) - 1);
                 if (charmap.get(s.charAt(left)) <= 0)
                 {
                    charmap.remove(s.charAt(left));
                 }
                left += 1;
            }

            longest = Math.max(longest, right - left + 1);
        }

        return longest;

    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        

        left = 0
        right = len(height) - 1

        maxleft = height[left]
        maxright = height[right]

        maxarea = 0

        while left < right:
            if maxleft <= maxright:
                left += 1
                maxleft = max(maxleft, height[left])
                maxarea += maxleft - height[left]
            else:
                right -= 1
                maxright = max(maxright, height[right])
                maxarea += maxright - height[right]
        
        return maxarea
